operator/cilium_node.go

```diff
@@ -1,112 +1,486 @@
+	"k8s.io/apimachinery/pkg/api/errors"
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/types"
 	"k8s.io/apimachinery/pkg/util/wait"
 	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
 )
 
 var (
 	// ciliumNodeStore contains all CiliumNodes present in k8s.
-	// Warning: The CiliumNodes stored in the cache are not intended to be
-	// used for Update operations in k8s as some of its fields are not
-	// populated.
 	ciliumNodeStore cache.Store
 
 	k8sCiliumNodesCacheSynced = make(chan struct{})
 )
 
-func startSynchronizingCiliumNodes(nodeManager allocator.NodeEventHandler) {
-	log.Info("Starting to synchronize CiliumNode custom resources")
+// ciliumNodeWatcherCfg contains the configuration for the CiliumNodeWatcher
+type ciliumNodeWatcherCfg struct {
+	resourceEventHandler  cache.ResourceEventHandler
+	ciliumNodeConvertFunc informer.ConvertFunc
+	withKVStore           bool
+	queue                 workqueue.RateLimitingInterface
+	syncHandler           func(key string) error
+}
 
-	// TODO: The operator is currently storing a full copy of the
-	// CiliumNode resource, as the resource grows, we may want to consider
-	// introducing a slim version of it.
-	var ciliumNodeInformer cache.Controller
-	ciliumNodeStore, ciliumNodeInformer = informer.NewInformer(
-		cache.NewListWatchFromClient(ciliumK8sClient.CiliumV2().RESTClient(),
-			v2.CNPluralName, v1.NamespaceAll, fields.Everything()),
-		&v2.CiliumNode{},
-		0,
-		cache.ResourceEventHandlerFuncs{
+// ciliumNodeWatcherCfgBuilder is used to generate a ciliumNodeWatcherCfg
+type ciliumNodeWatcherCfgBuilder struct {
+	nodeManager allocator.NodeEventHandler
+	kvStoreSync bool
+	cfgBuilt    atomic.Value
+}
+
+// NewCiliumNodeWatcherCfgBuilder returns a new ciliumNodeWatcherCfgBuilder.
+func NewCiliumNodeWatcherCfgBuilder() *ciliumNodeWatcherCfgBuilder {
+	var cfgBuilt atomic.Value
+	cfgBuilt.Store(false)
+	return &ciliumNodeWatcherCfgBuilder{
+		cfgBuilt: cfgBuilt,
+	}
+}
+
+// WithNodeManager sets the ciliumNodeWatcherCfgBuilder with a nodeManager that
+// will be used in the CiliumNode watch events.
+func (cfgBuilder *ciliumNodeWatcherCfgBuilder) WithNodeManager(nodeManager allocator.NodeEventHandler) *ciliumNodeWatcherCfgBuilder {
+	// We shouldn't change modify cfgBuilder.nodeManager after a
+	// ciliumNodeWatcherCfg was created with this ciliumNodeWatcherCfgBuilder
+	// because that field will be used to handle Kubernetes events.
+	if cfgBuilder.cfgBuilt.Load().(bool) {
+		panic("node manager already set")
+	}
+	cfgBuilder.nodeManager = nodeManager
+	return cfgBuilder
+}
+
+// WithKVStoreSync sets the ciliumNodeWatcherCfgBuilder to be configured
+// with the KVStoreSync support.
+func (cfgBuilder *ciliumNodeWatcherCfgBuilder) WithKVStoreSync() *ciliumNodeWatcherCfgBuilder {
+	cfgBuilder.kvStoreSync = true
+	return cfgBuilder
+}
+
+// Build builds and if WithKVStoreSync was called it will start a SharedStore.
```

Comment:
Build builds what? :joy: 
```suggestion
// Build builds the configuration and if WithKVStoreSync was called it will start a SharedStore.
```